### DISCUSSION TITLE
New Backend and now what

### DISCUSSION THREAD

**Discussion by benkb** [minoki/LunarML#12] [General]
Out of a spontaneous act I've forked LunarML and got started with adding a new Backend. After lots of hacking and grepping I finally got to a point where the whole thing compiles and runs. It is a copy of the Javascript backend and at this moment it still outputs javascript. 
I now have a little bit of a strange problem and I am a bit stuck. Since I really was not going along with the way the sml code in LunarML is formated I ran everything through smlfmt. I like the standard formatting of smlfmt much more and it fits a lot better with my Vim config. But I'm not sure what to do now, how I should proceed with git because of the reformatting I fear there are problems it my backend is maybe one day ready for being a pull request.

TLDR: I reformated everything with smlfmt, not sure how to proceed with git

**Comment by minoki** [comment]
I think that formatting with smlfmt is a good thing. Maybe it is a good time to format the entire code with smlfmt. Will that solve your problem?

(When I started the development of LunarML in 2018, smlfmt didn't exist and I was naively following sml-mode on Emacs.)

**Reply by benkb** [reply]
Through the hacking on my backend I got a deeper look into the LunarML codebase. There are parts that are (at least from my perspective) very complex. So I guess this project was very labor and time intensive and you must have spent a lot of time with that code. Therefore I very much appreciate you consider a reformatting.

My experiences with code formatters/prettifiers were not always good. (I have yet to find a good Scheme formatter that serves me well). But in the case of `smlfmt` I had a really good experience. The reformatting is reliable and also syntax errors are detected.

**Reply by benkb** [reply]
Like I said I've forked LunarML and created my own branch. If you want to have a look at it, heres the link: https://github.com/benkb/LunarML/tree/benkb-features

**Reply by minoki** [reply]
I've just reformatted the compiler code with smlfmt. I hope this will make it easier to merge my master to yours or rebase your fork on mine.

So you are interested in a Perl backend? I don't write Perl, but it would be cool if I could obtain a Perl script from SML code.
